### PR TITLE
pep257 cleanup of cookiecutter.exceptions

### PR DESCRIPTION
**Pull request by MinchinWeb**
cf #742

### CODE REVIEW

File: cookiecutter/exceptions.py [hunk 1]
```diff
@@ -1,103 +1,126 @@
...
+
     Raised when cookiecutter is called with both `no_input==True` and
     `replay==True` at the same time.
     """
 
 
 class FailedHookException(CookiecutterException):
     """
-    Raised when a hook script fails
+    Exemption for hook failures.
```
Comment (insspb) [inline cookiecutter/exceptions.py]: Exemption?

File: cookiecutter/exceptions.py [hunk 2]
```diff
@@ -1,103 +1,126 @@
...
+
+    Raised when a hook script fails.
     """
 
 
 class UndefinedVariableInTemplate(CookiecutterException):
-    """Raised when a template uses a variable which is not defined in the
+    """
+    Exemption for out-of-scope variables.
```
Comment (insspb) [inline cookiecutter/exceptions.py]: Exemption?

File: cookiecutter/exceptions.py [hunk 3]
```diff
@@ -1,103 +1,126 @@
...
+
+    Raised when a template uses a variable which is not defined in the
     context.
     """
+
     def __init__(self, message, error, context):
+        """Exemption for out-of-scope variables."""
```
Comment (insspb) [inline cookiecutter/exceptions.py]: Exemption?

File: cookiecutter/exceptions.py [hunk 4]
```diff
@@ -106,21 +129,33 @@ def __str__(self):
 
 
 class UnknownExtension(CookiecutterException):
-    """Raised when an environment is unable to import a required extension."""
+    """
+    Exemption for un-importable extention.
```
Comment (insspb) [inline cookiecutter/exceptions.py]: Exemption?

File: cookiecutter/exceptions.py [hunk 5]
```diff
@@ -106,21 +129,33 @@ def __str__(self):
...
+
+    Raised when an environment is unable to import a required extension.
+    """
 
 
 class RepositoryNotFound(CookiecutterException):
     """
+    Exemption for missing repo.
```
Comment (insspb) [inline cookiecutter/exceptions.py]: Exemption?

Comment (insspb) [reply cookiecutter/exceptions.py]: Exemption?

File: cookiecutter/exceptions.py [hunk 6]
```diff
@@ -106,21 +129,33 @@ def __str__(self):
...
+    """
+    Exemption for un-cloneable repo.
+
+    Raised when a cookiecutter template can't be cloned.
+    """
 
 
 class InvalidZipRepository(CookiecutterException):
     """
+    Exemption for bad zip repo.
```
Comment (insspb) [inline cookiecutter/exceptions.py]: Exemption?

**Comment by MinchinWeb** [comment]
@insspb Typo fixed!

`UnknownTemplateDirException` and `MissingProjectDir` aren't used locally, but I hesitate to remove them in case downstream projects make use of them. Their remove should at least be mentioned in the changelog...

Also, #1112 (from a year ago) covers much of this same ground.

**Pull request merged**